internal/test/framework/windowsvm.go

```diff
@@ -51,57 +52,114 @@ type TestWindowsVM interface {
+			instanceStatus := machine.Status
+			if instanceStatus.Phase == nil {
+				continue
+			}
+			instancePhase := *instanceStatus.Phase
+			if instancePhase == phaseProvisioned {
+				provisionedMachines = append(provisionedMachines, machine)
+			}
 		}
-		w.Windows = &types.Windows{}
-		w.Credentials = credentials
+		time.Sleep(5 * time.Second)
 	}
-
-	if err := w.SetupWinRMClient(); err != nil {
-		return w, fmt.Errorf("failed to setup winRM client for the Windows VM: %v", err)
+	if vmCount == len(provisionedMachines) {
+		return provisionedMachines, nil
 	}
-	// Wait for some time before starting configuring of ssh server. This is to let sshd service be available
-	// in the list of services
-	// TODO: Parse the output of the `Get-Service sshd, ssh-agent` on the Windows node to check if the windows nodes
-	// has those services present
-	if !skipSetup {
-		time.Sleep(time.Minute)
-		if err := w.ConfigureOpenSSHServer(); err != nil {
-			return w, fmt.Errorf("failed to configure OpenSSHServer on the Windows VM: %v", err)
-		}
+	return nil, fmt.Errorf("expected event count %d but got %d", vmCount, len(provisionedMachines))
+}
+
+// newWindowsVM creates and sets up a Windows VM in the cloud and returns the WindowsVM interface that can be used to
+// interact with the VM. If no error is returned then it is guaranteed that the VM was
+// created and can be interacted with.
+func (f *TestFramework) newWindowsVM(vmCount int) ([]TestWindowsVM, error) {
+	w := make([]TestWindowsVM, vmCount)
+	err := f.createMachineSet()
+	if err != nil {
+		return nil, fmt.Errorf("unable to create Windows MachineSet: %v", err)
 	}
-	if err := w.GetSSHClient(); err != nil {
-		return w, fmt.Errorf("failed to get ssh client for the Windows VM created: %v", err)
+
+	provisionedMachines, err := f.waitForMachines(vmCount)
+	if err != nil {
+		return nil, err
 	}
 
+	for i, machine := range provisionedMachines {
+		winVM := &testWindowsVM{
+			Windows: &types.Windows{},
+		}
+
+		ipAddress := ""
+		for _, address := range machine.Status.Addresses {
+			if address.Type == core.NodeInternalIP {
+				ipAddress = address.Address
+			}
+		}
+		if len(ipAddress) == 0 {
+			return nil, fmt.Errorf("no associated internal ip for machine: %s", machine.Name)
+		}
+
+		// Get the instance ID associated with the Windows machine.
+		providerID := *machine.Spec.ProviderID
+		if len(providerID) == 0 {
+			return nil, fmt.Errorf("no provider id associated with machine")
+		}
+		// Ex: aws:///us-east-1e/i-078285fdadccb2eaa. We always want the last entry which is the instanceID
+		providerTokens := strings.Split(providerID, "/")
+		instanceID := providerTokens[len(providerTokens)-1]
+		if len(instanceID) == 0 {
+			return nil, fmt.Errorf("empty instance id in provider id")
+		}
+		creds := types.NewCredentials(instanceID, ipAddress, types.Username)
+		winVM.Credentials = creds
+		log.Print("setting up ssh")
+		winVM.Credentials.SetSSHKey(f.Signer)
```

Comment:
we should have a nil or error check for the credentials we are setting up in this code block

Reply:
i don't think we need to check error for getters and setters.